Fix parsing unknown value in PerfData

EmTeedee opened the pull request:
'U' is neither a valid SI unit nor a valid prefix, so excluding it in the UOM part of the regex should not cause any issues.

The UOM regex now also accepts the `μ` prefix. Icinga normalizes values, but the regex seems to try to include them anyway.

If there is no numeric value, the pie chart can't show anything, so we test for that.

closes #700

nilmerg commented on library/Icingadb/Util/PerfData.php:
```diff
@@ -426,12 +426,15 @@ protected function parse()
 
         $matches = array();
-        if (preg_match('@^(-?(?:\d+)?(?:\.\d+)?)([a-zA-Z%°]{1,3})$@u', $parts[0], $matches)) {
+        if (preg_match('@^((?:U|(?:-?\d+)?(?:\.\d+)?))([a-zA-TV-Z%°μ]{1,3})$@u', $parts[0], $matches)) {
```
The `-?` can't be moved inside the first non-capture group. `.9kW` is a valid value. See https://github.com/Icinga/icingaweb2/issues/4690 for details.

What's `A-TV-Z` for? The regex currently doesn't match `'U'` anyway. And even if it does, the `U` is consumed early and the range can't consume it again.

The `μ` is not necessary. You already mentioned Icinga's normalization, so why do you think it is?

Please change the regex so that the following matches: `-.9kW`, `UkW`, `-1.9kW`, `.9kW`, `1.9kW`

nilmerg replied:
`U|` is not optional. In my tests this always consumes a leading `U`, no matter what follows. `-kW` isn't valid anyway and just `kW` as well. (makes no sense) And even if supplied this way, it's correctly identified as value (`-`) and uom (`kW`).

So I'm fine with the usage of `is_numeric` to decide whether to null the value again. But the `A-Z` range can be kept in combination with `U|` at the front.

I've also optimized it a bit: `@^(U|-?(?:\d+)?(?:\.\d+)?)([a-zA-Z%°]{1,3})$@u`

I've run unittests against this and I don't get any (new) errors. Or have you an example that doesn't work with this?

nilmerg commented on library/Icingadb/Util/PerfData.php:
```diff
@@ -426,12 +426,15 @@ protected function parse()
             $this->value = $matches[1];
         } else {
             $this->value = $parts[0];
         }
+        if ($this->value === 'U') {
```
An empty line please after control structures.

EmTeedee commented:
While testing perfdata values, I noticed that min/max get unexpectedly dropped when there is no warn/crit. Bug: https://github.com/Icinga/icinga2/issues/9634

EmTeedee commented:
The `U|` is not optional, but everything on the other side of the pipe (`-?(?:\d+)?(?:\.\d+)?`) is.
Since the UOM in the regex requires 1-3 characters, it will consume the U in this simple perfdata:
`test=U` (which should be indicating that the value of U is unknown and has _no_ unit).

It doesn't show up in the UI, but a quick test shows that the U ends up in the unit field of PerfData
The value is correctly set to null (because `is_numeric` doesn't accept an empty string).

For testing the dummy_text can be `example|test=U`

I added a line to print the result in the parse function (`printf("value: %s / unit: %s\n", $this->value ?? 'NULL', $this->unit ?? 'NULL');`) result:
![icinga_perfdata_unit_U](https://user-images.githubusercontent.com/195600/212963377-0fff3589-16b5-40df-84b8-79409df0393e.png)

Here is an example with `php -a`, note that the `U` is stuck in `matches[2]`, which gets assigned to `$this->unit`:
```
php > $perfdata="U";
php > $parts = explode(';', $perfdata); $matches = array(); preg_match('@^(U|-?(?:\d+)?(?:\.\d+)?)([a-zA-Z%°]{1,3})$@u', $parts[0], $matches);
php > var_dump($matches);
array(3) {
  [0]=>
  string(1) "U"
  [1]=>
  string(0) ""
  [2]=>
  string(1) "U"
}
```

I've perused the icinga2 source and found that perfdata is parsed by looking for the last `[.0-9]` in the value and treating everything after that as UOM and the front as value. Maybe that would be a better approach rather than the regex?
see https://github.com/Icinga/icinga2/blob/master/lib/base/perfdatavalue.cpp#L252-L258

I've also removed the offending commit.

nilmerg commented:
Ah, I didn't test just with an `U` :roll_eyes: 

> Maybe that would be a better approach rather than the regex?

The remaining regex has been proven correct in recent years. I'd rather not replace it just because of a minor enhancement ;)

It's fine now. But please remember to restore the `isset` call in `isVisualizable`.
